private_sharing/api_views.py

```diff
@@ -130,103 +124,154 @@ def post(self, request, format=None):
+
+    def post(self, request):
+        super(ProjectFileDirectUploadView, self).post(request)
+
+        key = get_upload_path(self.project.id_label,
+                              self.form.cleaned_data['filename'])
 
         data_file = ProjectDataFile(
-            user=project_member.member.user,
-            file=form.cleaned_data['data_file'],
-            metadata=form.cleaned_data['metadata'],
-            direct_sharing_project=project)
+            user=self.project_member.member.user,
+            file=key,
+            metadata=self.form.cleaned_data['metadata'],
+            direct_sharing_project=self.project)
 
         data_file.save()
 
-        return Response({'id': data_file.id}, status=status.HTTP_201_CREATED)
+        s3 = S3Connection(settings.AWS_ACCESS_KEY_ID,
+                          settings.AWS_SECRET_ACCESS_KEY)
 
+        url = s3.generate_url(
+            expires_in=SIX_HOURS,
+            method='PUT',
+            bucket=settings.AWS_STORAGE_BUCKET_NAME,
+            key=key)
 
-class ProjectFileDeleteView(ProjectAPIView, APIView):
-    # pylint: disable=redefined-builtin, unused-argument
-    def post(self, request, format=None):
-        project = DataRequestProject.objects.get(
-            master_access_token=self.request.auth.master_access_token)
+        return Response({
+            'id': data_file.id,
+            'url': url,
+        }, status=status.HTTP_201_CREATED)
 
-        form = DeleteDataFileForm(request.data)
 
-        if not form.is_valid():
-            return Response({'errors': form.errors},
-                            status=status.HTTP_400_BAD_REQUEST)
+class ProjectFileDirectUploadCompletionView(ProjectFormBaseView):
+    """
+    Initiate a direct upload to S3 for a project by pre-signing and returning
+    the URL.
+    """
 
-        try:
-            project_member = DataRequestProjectMember.objects.get(
-                project=project,
-                project_member_id=form.cleaned_data['project_member_id'])
-        except DataRequestProjectMember.DoesNotExist:
-            project_member = None
+    form_class = DirectUploadDataFileCompletionForm
 
-        if (not project_member or
-                not project_member.joined or
-                not project_member.authorized or
-                project_member.revoked):
-            return Response(
-                {
-                    'errors': {
-                        'project_member_id': 'project_member_id is invalid'
-                    }
-                },
-                status=status.HTTP_400_BAD_REQUEST)
-
-        file_id = form.cleaned_data['file_id']
-        file_basename = form.cleaned_data['file_basename']
-        all_files = form.cleaned_data['all_files']
+    def post(self, request):
+        super(ProjectFileDirectUploadCompletionView, self).post(request)
+
+
+class ProjectFileUploadView(ProjectFormBaseView):
+    """
+    A form for regular ProjectDataFile uploads.
```

Comment:
What's a "regular" ProjectDataFile upload? (I'm pretty sure I know what you mean -- but let's make some docstrings that give a bit more context for future code navigators. :) )